<a href="https://colab.research.google.com/github/Alexandremsn/desafio_alura_data_science/blob/main/DesafioSemana1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Atividades:

Entender quais informações o conjunto de dados possui;

Analisar quais os tipos de dados;

Verificar quais são as inconsistências nos dados;

Corrigir as inconsistências nos dados;

Traduzir as colunas;

Criar coluna de contas diárias.


### Entender quais informações o conjunto de dados possui;


Importando as Bibiotecas

In [1]:
import pandas as pd

Importando os dados

In [2]:
dataset = pd.read_json("https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json")

varificando se a importação foi bem sucedida

In [3]:
dataset.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

Verificamos que os dados possuem diversas informações dos clientes seu perfil demográfico, seus serviços contratados, e seu gasto mensal.

###Analisar quais os tipos de dados

verificamos que os dados estão agrupado em dicionários assim vamos aplicar a normallização as variáveis que estão agrupadas normalizando cada coluna individualmente e verificando sua saída.

In [4]:
dataset_customer = pd.json_normalize(data=dataset.customer)
dataset_customer

gender  SeniorCitizen Partner Dependents  tenure
0     Female              0     Yes        Yes       9
1       Male              0      No         No       9
2       Male              0      No         No       4
3       Male              1     Yes         No      13
4     Female              1     Yes         No       3
...      ...            ...     ...        ...     ...
7262  Female              0      No         No      13
7263    Male              0     Yes         No      22
7264    Male              0      No         No       2
7265    Male              0     Yes        Yes      67
7266    Male              0     Yes        Yes      63

[7267 rows x 5 columns]

Agora os dados estão normalizados em outro dataframe vamos repetir o processo para as outras colunas

In [5]:
dataset_phone = pd.json_normalize(data=dataset.phone)
dataset_phone

PhoneService     MultipleLines
0             Yes                No
1             Yes               Yes
2             Yes                No
3             Yes                No
4             Yes                No
...           ...               ...
7262          Yes                No
7263          Yes               Yes
7264          Yes                No
7265          Yes                No
7266           No  No phone service

[7267 rows x 2 columns]

Normalizando dados para a variável internet

In [6]:
dataset_internet = pd.json_normalize(data=dataset.internet)
dataset_internet

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0                DSL             No          Yes               No         Yes   
1                DSL             No           No               No          No   
2        Fiber optic             No           No              Yes          No   
3        Fiber optic             No          Yes              Yes          No   
4        Fiber optic             No           No               No         Yes   
...              ...            ...          ...              ...         ...   
7262             DSL            Yes           No               No         Yes   
7263     Fiber optic             No           No               No          No   
7264             DSL             No          Yes               No          No   
7265             DSL            Yes           No              Yes         Yes   
7266             DSL            Yes          Yes              Yes          No   

     StreamingTV StreamingMovies  
0            Yes              No  
1             No             Yes  
2             No              No  
3            Yes             Yes  
4            Yes              No  
...          ...             ...  
7262          No              No  
7263          No             Yes  
7264          No              No  
7265          No             Yes  
7266         Yes             Yes  

[7267 rows x 7 columns]

Normalizando a varível account

In [7]:
dataset_account = pd.json_normalize(data=dataset.account)
dataset_account

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0           One year              Yes      Mailed check            65.60   
1     Month-to-month               No      Mailed check            59.90   
2     Month-to-month              Yes  Electronic check            73.90   
3     Month-to-month              Yes  Electronic check            98.00   
4     Month-to-month              Yes      Mailed check            83.90   
...              ...              ...               ...              ...   
7262        One year               No      Mailed check            55.15   
7263  Month-to-month              Yes  Electronic check            85.10   
7264  Month-to-month              Yes      Mailed check            50.30   
7265        Two year               No      Mailed check            67.85   
7266        Two year               No  Electronic check            59.00   

     Charges.Total  
0            593.3  
1            542.4  
2           280.85  
3          1237.85  
4            267.4  
...            ...  
7262         742.9  
7263        1873.7  
7264         92.75  
7265       4627.65  
7266        3707.6  

[7267 rows x 5 columns]

Juntando as variáveis normalizadas no dataset principal

In [8]:
dataset = pd.concat([dataset, dataset_customer,dataset_phone, dataset_internet, dataset_account], axis=1)
dataset.head(10)

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
5  0013-MHZWF    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
6  0013-SMEOE    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
7  0014-BMAQU    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
8  0015-UOCOJ    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
9  0016-QLJIS    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
5   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
6   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
8   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
9  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   
5  {'InternetService': 'DSL', 'OnlineSecurity': '...   
6  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7  {'InternetService': 'Fiber optic', 'OnlineSecu...   
8  {'InternetService': 'DSL', 'OnlineSecurity': '...   
9  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                             account  gender  SeniorCitizen  \
0  {'Contract': 'One year', 'PaperlessBilling': '...  Female              0   
1  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male              0   
2  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male              0   
3  {'Contract': 'Month-to-month', 'PaperlessBilli...    Male              1   
4  {'Contract': 'Month-to-month', 'PaperlessBilli...  Female              1   
5  {'Contract': 'Month-to-month', 'PaperlessBilli...  Female              0   
6  {'Contract': 'Two year', 'PaperlessBilling': '...  Female              1   
7  {'Contract': 'Two year', 'PaperlessBilling': '...    Male              0   
8  {'Contract': 'Month-to-month', 'PaperlessBilli...  Female              1   
9  {'Contract': 'Two year', 'PaperlessBilling': '...  Female              0   

  Partner Dependents  ...  OnlineBackup DeviceProtection TechSupport  \
0     Yes        Yes  ...           Yes               No         Yes   
1      No         No  ...            No               No          No   
2      No         No  ...            No              Yes          No   
3     Yes         No  ...           Yes              Yes          No   
4     Yes         No  ...            No               No         Yes   
5      No        Yes  ...            No               No         Yes   
6     Yes         No  ...           Yes              Yes         Yes   
7     Yes         No  ...            No               No         Yes   
8      No         No  ...            No               No          No   
9     Yes        Yes  ...           Yes              Yes         Yes   

  StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes              No        One year              Yes   
1          No             Yes  Month-to-month               No   
2          No              No  Month-to-month              Yes   
3         Yes             Yes  Month-to-month    

procedemos a exclusão das colunas fonte de dados
Alguns métodos do pandas podem apresentar erro se forem aplicadas em um coluna com '.' no nome da coluna, assim vamos renomear para evitar problemas


In [9]:
del dataset['customer']
del dataset['phone']
del dataset['internet']
del dataset['account']
dataset.rename(columns={'Charges.Monthly' : 'Charges_Monthly', 'Charges.Total' : 'Charges_Total'}, inplace = True)
dataset.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges_Monthly  Charges_Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

O dataset foi inteiramente normalizado e agora podemos ver com clareza os dados apresentados, o dataset é predominantemente qualitativo, tendo aglumas variáveis quantitativas. 

###Verificar quais são as inconsistências nos dados;

In [10]:
dataset.shape

(7267, 21)

Verificamos o tamanho da base de dados em linhas e colunas e verificamos se não possuimos campos com null

In [11]:
dataset.isna().sum()

customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges_Monthly     0
Charges_Total       0
dtype: int64

Não verificamos inconsistências deste tipo na base

In [12]:
dataset.Churn.value_counts()
# foram verificados valores vazios nesta variável
#foi continuada a verifição para os parâmetros:
#Dependents, Partner, SeniorCitizen, Gender, MultipleLines, PhoneService, DeviceProtection, InternetService,
#OnlineBackup, OnlineSecurity, StreamingTV, TechSupport, Contract, PaperlessBilling, PaymentMethod.
# comparamos os valores de pessoas que não possuem serviço de internet para verificar se o número é o mesmo indicado em todas as variáveis que listam este valor.
# para as variáveis a seguir devido ao grande range de valores possíveis foi decidido verificar individualmente cada uma delas:
## Tenure, Charges_Monthly, Charges_Total 

No     5174
Yes    1869
        224
Name: Churn, dtype: int64

Verificamos uma inconsitência temos o valor de um espaço (' ') em diversos campos.

In [13]:
print(dataset.gender.value_counts())
print(dataset.SeniorCitizen.value_counts())
print(dataset.Partner.value_counts())



Male      3675
Female    3592
Name: gender, dtype: int64
0    6085
1    1182
Name: SeniorCitizen, dtype: int64
No     3749
Yes    3518
Name: Partner, dtype: int64


analisamos as 3 variáveis ('gender', 'SeniorCitizen', 'Partner',), e não verificamos a pesença de valores anormais, verificamos que a variável 'SeniorCitizen', está em um formado 0 e 1 ao invez de yes, no como nas outras variáveis, então precisaremos realizar a substituição dos valores para manter a padronização.

In [14]:
print(dataset.Dependents.value_counts())
print(dataset.tenure.value_counts())
print(dataset.PhoneService.value_counts())

No     5086
Yes    2181
Name: Dependents, dtype: int64
1     634
72    369
2     246
3     207
4     185
     ... 
38     60
39     59
44     54
36     50
0      11
Name: tenure, Length: 73, dtype: int64
Yes    6560
No      707
Name: PhoneService, dtype: int64


verificamos que não temos muitas inconsistências, para as 3 variáveis analizadas ('Dependents', 'tenure', 'PhoneService'), vamos verificar na variável 'tenure' os valores máximos e mínimos para verificar se não possui a correência de valores irreais.

In [15]:
print('o valor máximo da variável é:',(max(dataset.tenure)))
print('o valor minimo da variável é:',(min(dataset.tenure)))

o valor máximo da variável é: 72
o valor minimo da variável é: 0


os valores de 0 a 72 meses para tempo de contrato parecem dentro do esperado.
a varíavel não possui este tipo de inconsistência.

In [16]:
print(dataset.MultipleLines.value_counts())
print(dataset.InternetService.value_counts())
print(dataset.OnlineSecurity.value_counts())

No                  3495
Yes                 3065
No phone service     707
Name: MultipleLines, dtype: int64
Fiber optic    3198
DSL            2488
No             1581
Name: InternetService, dtype: int64
No                     3608
Yes                    2078
No internet service    1581
Name: OnlineSecurity, dtype: int64


Temos valores dentro do esperado aproveitamos para comparar o valor de 1581, de clientes sem serviço de internet com as outras veriáveis que também exibem este quantitativo.

In [17]:
print(dataset.OnlineBackup.value_counts())
print(dataset.DeviceProtection.value_counts())
print(dataset.TechSupport.value_counts())

No                     3182
Yes                    2504
No internet service    1581
Name: OnlineBackup, dtype: int64
No                     3195
Yes                    2491
No internet service    1581
Name: DeviceProtection, dtype: int64
No                     3582
Yes                    2104
No internet service    1581
Name: TechSupport, dtype: int64


dentro do esperado não temos valores fora de padrão o valor de sen internet também está o mesmo.

In [18]:
print(dataset.StreamingTV.value_counts())
print(dataset.StreamingMovies.value_counts())
print(dataset.Contract.value_counts())

No                     2896
Yes                    2790
No internet service    1581
Name: StreamingTV, dtype: int64
No                     2870
Yes                    2816
No internet service    1581
Name: StreamingMovies, dtype: int64
Month-to-month    4005
Two year          1743
One year          1519
Name: Contract, dtype: int64


Valores dentro do esperado, na próxima análise temos variáveis quantitativas assim vamos analizar o máximo e mínimo para verificar se são temos valores absurdamete grandes ou negativos.

In [19]:
print('o valor máximo da variável Charges_Monthly é:',(max(dataset.Charges_Monthly)))
print('o valor minimo da variável Charges_Monthly é:',(min(dataset.Charges_Monthly)))
print('o valor máximo da variável Charges_Total é:',(max(dataset.Charges_Total)))
print('o valor minimo da variável Charges_Total é:',(min(dataset.Charges_Total)))

o valor máximo da variável Charges_Monthly é: 118.75
o valor minimo da variável Charges_Monthly é: 18.25
o valor máximo da variável Charges_Total é: 999.9
o valor minimo da variável Charges_Total é:  


Verificamos os valoes estão em uma faixa ok, porém no valor mímimo vemos uma inconsitência que pode indicar que a coluna possui um tipo de dado diferente.
abaixo vamos verificar os tipos de dados.

In [20]:
result = dataset['tenure'].dtypes
result2 = dataset['Charges_Monthly'].dtypes
result3 = dataset['Charges_Total'].dtypes

print("Output, tenure:")
print(result)
print("Output, Charges_Monthly:")
print(result2)
print("Output, Charges_Total:")
print(result3)



Output, tenure:
int64
Output, Charges_Monthly:
float64
Output, Charges_Total:
object


verficamos um probelma a Charges_total está em um tipo diferente vamos precisar converter o tipo.

###Corrigir as inconsistências nos dados;

As inconsitências detectadas são:
Churn com valores inválidos ' ';
SeniorCitzen com valores fora de padâo;
Charges_Total com dados de outro tipo que não numérico

trocamos os valores inválidos por 0, porém vamos verificar se este espaço vazio é aunsência de dados por erro ou deve ser subtituido por não.

In [21]:
for i in range(dataset.shape[0]):
  if (dataset.Charges_Total[i] == " "):
    dataset.Charges_Total[i] = 0
    

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


recebemos a infromação que os valores vazios significavam ausência de dados, assim vamos verificar o percentual destas inconsistências. para verificar o impacto da exclusão destas linhas na base de dados.

In [22]:
dataset.Churn.value_counts(normalize = True)*100

No     71.198569
Yes    25.719004
        3.082427
Name: Churn, dtype: float64

Verificamos que 3% dos dados estão sem valor, vamos remover os valores da base pois o valor será o foco de nossa análise.

In [23]:
print(dataset.shape)
for i in range(dataset.shape[0]):
  valores = ['No','Yes']
  if (not(dataset.Churn[i] in valores)):
    dataset.drop(i,axis =0, inplace=True)
    
dataset.reset_index(drop=True, inplace=True)   

print(dataset.shape)

(7267, 21)
(7043, 21)


In [24]:
dataset.Churn.value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

Realizamos a exclusão dos valores e verificamos que agora temos na varível apenas os valores Yes e No.

Agora vamos corrigir a inconsitência da coluna 'Charges_Total', que possuiam dados do tipo string, vamos converte-los para número

In [25]:
dataset['Charges_Total'] = pd.to_numeric(dataset['Charges_Total'])
dataset.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges_Monthly  Charges_Total  
0              Yes      Mailed check            65.6         593.30  
1               No      Mailed check            59.9         542.40  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9         267.40  

[5 rows x 21 columns]

In [26]:
dataset.Charges_Total.dtypes

dtype('float64')

Sucesso, agora a coluna possui apenas dados do tipo numérico.

Agora vamos padornizar os dados da variável 'SeniorCitizen', trocando o o por No e o 1 por Yes.

In [27]:
dataset['SeniorCitizen'] = dataset['SeniorCitizen'].replace([0,1],['No','Yes'])

dataset.head(10)

customerID Churn  gender SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female            No     Yes        Yes       9   
1  0003-MKNFE    No    Male            No      No         No       9   
2  0004-TLHLJ   Yes    Male            No      No         No       4   
3  0011-IGKFF   Yes    Male           Yes     Yes         No      13   
4  0013-EXCHZ   Yes  Female           Yes     Yes         No       3   
5  0013-MHZWF    No  Female            No      No        Yes       9   
6  0013-SMEOE    No  Female           Yes     Yes         No      71   
7  0014-BMAQU    No    Male            No     Yes         No      63   
8  0015-UOCOJ    No  Female           Yes      No         No       7   
9  0016-QLJIS    No  Female            No     Yes        Yes      65   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   
5          Yes            No             DSL  ...           No   
6          Yes            No     Fiber optic  ...          Yes   
7          Yes           Yes     Fiber optic  ...           No   
8          Yes            No             DSL  ...           No   
9          Yes           Yes             DSL  ...          Yes   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   
5               No         Yes         Yes             Yes  Month-to-month   
6              Yes         Yes         Yes             Yes        Two year   
7               No         Yes          No              No        Two year   
8               No          No          No              No  Month-to-month   
9              Yes         Yes         Yes             Yes        Two year   

  PaperlessBilling              PaymentMethod Charges_Monthly  Charges_Total  
0              Yes               Mailed check           65.60         593.30  
1               No               Mailed check           59.90         542.40  
2              Yes           Electronic check           73.90         280.85  
3              Yes           Electronic check           98.00        1237.85  
4              Yes               Mailed check           83.90         267.40  
5              Yes    Credit card (automatic)           69.40         571.45  
6              Yes  Bank transfer (automatic)          109.70        7904.25  
7              Yes    Credit card (automatic)           84.65        5377.80  
8              Yes           Electronic check           48.20         340.35  
9              Yes               Mailed check           90.45        5957.90  

[10 rows x 21 columns]

Agora as inconsitências detectadas foram corrigidas.

###Traduzir as colunas

In [28]:
dataset.columns

Index(['customerID', 'Churn', 'gender', 'SeniorCitizen', 'Partner',
       'Dependents', 'tenure', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'Charges_Monthly',
       'Charges_Total'],
      dtype='object')

In [29]:
dataset.rename(columns={'customerID': 'ID_Cliente', 'Churn' : 'Churn', 'Dependents' : 'Dependentes', 'Partner': 'Parceiro', 'SeniorCitizen': 'Maior_65',
       'gender' : 'Genero', 'tenure' : 'Meses_de_Contrato', 'MultipleLines' : 'Mais_de_uma_linha', 'PhoneService' : 'Serviço_telefone', 'DeviceProtection' : 'Proteção_de_dispositivo',
       'InternetService' : 'Serviço_Internet', 'OnlineBackup': 'Backup_Online', 'OnlineSecurity' : 'Seguraca_Online', 'StreamingMovies' : 'Streaming_Filmes',
       'StreamingTV' :'Streaming_TV' , 'TechSupport' : 'Suporte_Tecnico', 'Charges_Monthly' : 'Gasto_Mensal', 'Charges_Total' : 'Gasto_Total',
       'Contract' : 'Tipo_Contrato', 'PaperlessBilling' : 'Conta_Online', 'PaymentMethod' :  'Metodo_de_Pagamento'}, inplace = True)

aplicamos este código para subistituir os índices pelo seu equivalente em portugês.

In [30]:
dataset.head()

ID_Cliente Churn  Genero Maior_65 Parceiro Dependentes  Meses_de_Contrato  \
0  0002-ORFBO    No  Female       No      Yes         Yes                  9   
1  0003-MKNFE    No    Male       No       No          No                  9   
2  0004-TLHLJ   Yes    Male       No       No          No                  4   
3  0011-IGKFF   Yes    Male      Yes      Yes          No                 13   
4  0013-EXCHZ   Yes  Female      Yes      Yes          No                  3   

  Serviço_telefone Mais_de_uma_linha Serviço_Internet  ... Backup_Online  \
0              Yes                No              DSL  ...           Yes   
1              Yes               Yes              DSL  ...            No   
2              Yes                No      Fiber optic  ...            No   
3              Yes                No      Fiber optic  ...           Yes   
4              Yes                No      Fiber optic  ...            No   

  Proteção_de_dispositivo Suporte_Tecnico Streaming_TV Streaming_Filmes  \
0                      No             Yes          Yes               No   
1                      No              No           No              Yes   
2                     Yes              No           No               No   
3                     Yes              No          Yes              Yes   
4                      No             Yes          Yes               No   

    Tipo_Contrato Conta_Online Metodo_de_Pagamento Gasto_Mensal  Gasto_Total  
0        One year          Yes        Mailed check         65.6       593.30  
1  Month-to-month           No        Mailed check         59.9       542.40  
2  Month-to-month          Yes    Electronic check         73.9       280.85  
3  Month-to-month          Yes    Electronic check         98.0      1237.85  
4  Month-to-month          Yes        Mailed check         83.9       267.40  

[5 rows x 21 columns]

verificamos que os indíces foram corretamente substituídos.

###Criar coluna de contas diárias

para criar a coluna calculada do gasto diário, dividimos o gasto mensal por 30, como o valor é monetário, aplicamos o aredondamento para duas casas decimais. 

In [31]:
dataset['Gasto_Diario'] = (round((dataset['Gasto_Mensal']/30),2))
dataset.head(1)

ID_Cliente Churn  Genero Maior_65 Parceiro Dependentes  Meses_de_Contrato  \
0  0002-ORFBO    No  Female       No      Yes         Yes                  9   

  Serviço_telefone Mais_de_uma_linha Serviço_Internet  ...  \
0              Yes                No              DSL  ...   

  Proteção_de_dispositivo Suporte_Tecnico Streaming_TV Streaming_Filmes  \
0                      No             Yes          Yes               No   

  Tipo_Contrato Conta_Online Metodo_de_Pagamento Gasto_Mensal Gasto_Total  \
0      One year          Yes        Mailed check         65.6       593.3   

   Gasto_Diario  
0          2.19  

[1 rows x 22 columns]

In [32]:
dataset.columns

Index(['ID_Cliente', 'Churn', 'Genero', 'Maior_65', 'Parceiro', 'Dependentes',
       'Meses_de_Contrato', 'Serviço_telefone', 'Mais_de_uma_linha',
       'Serviço_Internet', 'Seguraca_Online', 'Backup_Online',
       'Proteção_de_dispositivo', 'Suporte_Tecnico', 'Streaming_TV',
       'Streaming_Filmes', 'Tipo_Contrato', 'Conta_Online',
       'Metodo_de_Pagamento', 'Gasto_Mensal', 'Gasto_Total', 'Gasto_Diario'],
      dtype='object')

In [33]:
columnsorder = ['ID_Cliente', 'Churn', 'Genero', 'Maior_65', 'Parceiro', 'Dependentes',
       'Meses_de_Contrato', 'Serviço_telefone', 'Mais_de_uma_linha', 'Serviço_Internet',
       'Seguraca_Online', 'Backup_Online', 'Proteção_de_dispositivo',
       'Suporte_Tecnico', 'Streaming_TV', 'Streaming_Filmes', 'Tipo_Contrato',
       'Conta_Online', 'Metodo_de_Pagamento', 'Gasto_Diario', 'Gasto_Mensal', 'Gasto_Total'
       ]

agora vamos reordenar as colunas pois como nova coluna foi gerada no final vamos reordenar para que faça mais sentido. para que tenha a gradação diário, mensal e total.

In [34]:
dataset = dataset[columnsorder]
dataset.head(5)

ID_Cliente Churn  Genero Maior_65 Parceiro Dependentes  Meses_de_Contrato  \
0  0002-ORFBO    No  Female       No      Yes         Yes                  9   
1  0003-MKNFE    No    Male       No       No          No                  9   
2  0004-TLHLJ   Yes    Male       No       No          No                  4   
3  0011-IGKFF   Yes    Male      Yes      Yes          No                 13   
4  0013-EXCHZ   Yes  Female      Yes      Yes          No                  3   

  Serviço_telefone Mais_de_uma_linha Serviço_Internet  ...  \
0              Yes                No              DSL  ...   
1              Yes               Yes              DSL  ...   
2              Yes                No      Fiber optic  ...   
3              Yes                No      Fiber optic  ...   
4              Yes                No      Fiber optic  ...   

  Proteção_de_dispositivo Suporte_Tecnico Streaming_TV Streaming_Filmes  \
0                      No             Yes          Yes               No   
1                      No              No           No              Yes   
2                     Yes              No           No               No   
3                     Yes              No          Yes              Yes   
4                      No             Yes          Yes               No   

    Tipo_Contrato Conta_Online Metodo_de_Pagamento Gasto_Diario Gasto_Mensal  \
0        One year          Yes        Mailed check         2.19         65.6   
1  Month-to-month           No        Mailed check         2.00         59.9   
2  Month-to-month          Yes    Electronic check         2.46         73.9   
3  Month-to-month          Yes    Electronic check         3.27         98.0   
4  Month-to-month          Yes        Mailed check         2.80         83.9   

   Gasto_Total  
0       593.30  
1       542.40  
2       280.85  
3      1237.85  
4       267.40  

[5 rows x 22 columns]

In [35]:
dataset.to_csv('dados_normalizados_churn.csv', index=False)

Finalizamos exportando a base para utiliza-la na semana seguinte.